### PR TITLE
[BEAM-14557] Read and Seek Runner Capabilities in Go SDK 

### LINKED ISSUE #21807: [Task]: Go SDK should use runner capabilities to decide whether to send monitoring info with short ID
### What needs to happen?

Read the runner capabilities in boot.go and plumb it to the SDK harness so that harness can take decide whether to send monitoring infos with short ID only.

### Issue Priority

Priority: 2

### Issue Component

Component: sdk-go

### PR DESCRIPTION
**Pull request by riteshghorse**
Closes #21807
This PR provides parity with Java and Python SDK to read the runner capabilities and use it to optimize/support additional features/utilities. This PR focuses on supporting monitoring info with short ID.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Add a link to the appropriate issue in your description, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by riteshghorse** [comment]
R: @lostluck

**Review by lostluck — APPROVED** [review]
Woohoo! LGTM

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by riteshghorse** [comment]
Okay, it looks like the universal runner extracts metrics from MonitoringInfos for producing the pipeline result. Need to change that as well.

**Comment by riteshghorse** [comment]
The portable and flink runners are not requesting `req.MonitoringInfos()`. That's the reason they are not getting any monitoring infos with the new approach.

**Comment by lostluck** [comment]
As discussed, we can't remove the monitoring infos due to other runners, but I have found the correct URNs we need to look out for and probably provide as a capability:

"beam:protocol:monitoring_info_short_ids:v1"

https://source.corp.google.com/piper///depot/google3/third_party/apache_beam/model/pipeline/org/apache/beam/model/pipeline/v1/beam_runner_api.proto;l=1639?q=StandardRunnerProtocols&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental

**Review by lostluck — CHANGES_REQUESTED** [review]

File: sdks/go/pkg/beam/core/runtime/graphx/translate.go [hunk 1]
```diff
@@ -86,7 +86,7 @@ const (
 
 func goCapabilities() []string {
 	capabilities := []string{
-		URNLegacyProgressReporting,
+		URNProgressReporting,
```
[the remark this answers is not in the document]
Comment (riteshghorse) [reply sdks/go/pkg/beam/core/runtime/graphx/translate.go:90]: Done.

**Comment by lostluck** [comment]
Since the scope of the PR has changed, could the Title and PR Description be updated accordingly?

File: sdks/go/pkg/beam/core/runtime/harness/harness.go [hunk 2]
```diff
@@ -371,6 +390,10 @@ func (c *control) handleInstruction(ctx context.Context, req *fnpb.InstructionRe
 		c.cache.CompleteBundle(tokens...)
 
 		mons, pylds := monitoring(plan, store)
+		if c.runnerCapabilities[URNMonitoringInfoShortID] {
+			mons = []*pipeline_v1.MonitoringInfo{}
+		}
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/harness/harness.go:395]: So this only accomplishes half the goal of the short-ids approach to monitoring. While we are no longer sending the larger payloads, we are still allocating and authoring them, which we should still avoid doing if we can avoid it. The monitoring method is what needs changing to "understand" the capabilities and provide one thing or another (or if we have no information, both).

Comment (riteshghorse) [reply sdks/go/pkg/beam/core/runtime/harness/harness.go]: yeah, that is taking up memory as well as space. Fixed.

**Comment by lostluck** [comment]
retest this please

**Comment by lostluck** [comment]
https://ci-beam.apache.org/job/beam_PostCommit_Go_PR/438/ passed! Merging this.

**Review by lostluck — APPROVED** [review]
LGTM

**Pull request merged**